File: examples/comparative/shakuntala_llm.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
================================================================================
Kalidasa's Abhijnanasakuntalam Autoregressive Language Model (Python Reference)
महाकविकालिदासविरचितस्य अभिज्ञानशाकुन्तलस्य भाषा-प्रतिरूपम् (पायथन-तुलना)
================================================================================

This program is the official, well-documented Python equivalent of the pure
Devanagari Sankode implementation found in `examples/शाकुन्तल_सम्पूर्ण_प्रतिरूप.सङ्`.

It provides a line-by-line comparative reference to illustrate how:
1. Object-oriented data structures in Python correspond to Sankode's `संरचना` & `विधान`.
2. Python's dynamic runtime and garbage collection compare with Sankode's
   compile-time affine ownership (`स्वामित्व`), move semantics, and borrowing (`ऋण`).
3. Mathematical modeling (Laplace smoothing, cross-entropy loss, perplexity, and
   temperature-scaled softmax autoregression) is expressed across both languages.
4. File I/O and weight serialization operate equivalently.
"""

from __future__ import annotations
import math
import os
import sys
from typing import List

# Ensure UTF-8 output encoding across all operating systems
if sys.stdout.encoding != "utf-8":
    try:
        sys.stdout.reconfigure(encoding="utf-8")
    except AttributeError:
        pass


# ==============================================================================
# 1. Pseudo-Random Number Generator (यादृच्छिक-संख्या-यन्त्रम्)
# ==============================================================================
class RandomGenerator:
    """
    A 31-bit Linear Congruential Generator (LCG) replicating Sankode's PRNG.

    Formula:
        X_{n+1} = (a * X_n + c) mod m
        where a = 1664525, c = 1013904223, m = 2^31 - 1 = 2147483647

    Comparative Note:
        - In Sankode: Implemented via `संरचना यादृच्छिक_यन्त्र` and `विधान`.
          The mutable receiver `चलऋण स्व` (mutable borrow) guarantees that only
          one thread or caller mutates the PRNG seed at any point in time.
        - In Python: A standard class with mutable state `self.seed`.
    """

    def __init__(self, seed: int = 987654321) -> None:
        self.seed: int = seed

    def next_int(self, limit: int) -> int:
        """Generates a pseudo-random integer in [0, limit)."""
        if limit <= 1:
            return 0
        self.seed = (self.seed * 1664525 + 1013904223) % 2147483647
        result = self.seed % limit
        return abs(result)

    def random_float(self) -> float:
        """Generates a pseudo-random floating point number in [0.0, 1.0)."""
        val = self.next_int(10000)
        return float(val) / 10000.0


# ==============================================================================
# 2. Vocabulary & Character Tokenizer (शब्दावली)
# ==============================================================================
class Vocabulary:
    """
    Character-level tokenizer and bidirectional vocabulary indexer.

    Comparative Note:
        - In Sankode: Represents `संरचना शब्दावली` holding `वर्णाः: सूची` (dynamic list).
          String slices and characters are manipulated via native Devanagari intrinsics:
          `सूत्र_दैर्घ्यम्`, `सूत्र_वर्ण`, and `सूत्र_विभाजय`.
        - In Python: Uses a standard `list[str]` and dict for O(1) index lookups.
    """

    def __init__(self, tokens: List[str] | None = None) -> None:
        self.tokens: List[str] = [] if tokens is None else list(tokens)
        self.token_to_idx: dict[str, int] = {t: i for i, t in enumerate(self.tokens)}

    def add(self, char: str) -> int:
        """Adds a new character token to the vocabulary if not already present."""
        if char not in self.token_to_idx:
            idx = len(self.tokens)
            self.tokens.append(char)
            self.token_to_idx[char] = idx
            return idx
        return self.token_to_idx[char]

    def find(self, char: str) -> int:
        """Returns the integer index of the character, or -1 if unseen."""
        return self.token_to_idx.get(char, -1)

    def get(self, idx: int) -> str:
        """Retrieves the character token at index `idx`."""
        return self.tokens[idx]

    @property
    def size(self) -> int:
        """Total number of unique tokens in vocabulary (|V|)."""
        return len(self.tokens)


# ==============================================================================
# 3. Autoregressive Language Model (भाषा-प्रतिरूपम्)
# ==============================================================================
class LanguageModel:
    """
    Autoregressive bigram language model with Laplace smoothing and temperature sampling.

    Mathematical Formulation:
        1. Transition Probability Matrix P of shape |V| x |V|:
           P(w_t | w_{t-1}) = (Count(w_{t-1}, w_t) + 1.0) / (Count(w_{t-1}) + |V|)
           where add-1.0 is Laplace smoothing preventing zero-probability bottlenecks.
        2. Cross-Entropy Loss (Negative Log-Likelihood):
           L = - (1 / N) * sum_{t=1}^{N} ln P(w_t | w_{t-1})
        3. Perplexity:
           PPL = exp(L)
        4. Softmax Sampling with Temperature T:
           P_T(w_t | w_{t-1}) propto exp( ln P(w_t | w_{t-1}) / T )

    Comparative Note:
        - In Sankode:
          `संरचना भाषा_प्रतिरूप` stores `भार: सूची` (flattened 1D array of size |V|*|V|).
          Compile-time ownership ensures that `भार` cannot be accessed after being
          moved, and methods borrow the model via `चलऋण स्व` (mutable) or `ऋण स्व` (read-only).
        - In Python:
          Stores weights in a 1D `list[float]` to maintain exact memory-layout parity.
    """

    def __init__(self, vocab: Vocabulary, weights: List[float] | None = None) -> None:
        self.vocab: Vocabulary = vocab
        self.size: int = vocab.size
        total_elements = self.size * self.size
        if weights is not None:
            self.weights: List[float] = weights
        else:
            self.weights = [0.0] * total_elements

    def train(self, text: str) -> None:
        """
        Trains bigram transition counts on the input text corpus using Laplace add-one smoothing.
        """
        v = self.size
        counts = [0.0] * (v * v)
        row_sums = [0.0] * v

        # Count bigram transitions C(w_prev, w_next)
        text_len = len(text)
        for i in range(text_len - 1):
            c_prev = text[i]
            c_next = text[i + 1]
            u = self.vocab.find(c_prev)
            w = self.vocab.find(c_next)
            if u >= 0 and w >= 0:
                counts[u * v + w] += 1.0
                row_sums[u] += 1.0

        # Apply Laplace add-one smoothing & normalize rows to probabilities
        for r in range(v):
            denom = row_sums[r] + float(v)
            for c in range(v):
                idx = r * v + c
                self.weights[idx] = (counts[idx] + 1.0) / denom

    def compute_loss(self, text: str) -> float:
        """
        Computes negative log-likelihood cross-entropy loss over the corpus text.
        """
        v = self.size
        text_len = len(text)
        if text_len <= 1:
            return 0.0

        total_nll = 0.0
        valid_transitions = 0.0

        for i in range(text_len - 1):
            c_prev = text[i]
            c_next = text[i + 1]
            u = self.vocab.find(c_prev)
            w = self.vocab.find(c_next)
            if u >= 0 and w >= 0:
                prob = self.weights[u * v + w]
                if prob > 0.0:
                    total_nll -= math.log(prob)
                    valid_transitions += 1.0

        return total_nll / valid_transitions if valid_transitions > 0 else 0.0

    def save(self, filepath: str) -> None:
        """
        Serializes model vocabulary and transition weights to disk.
        Uses exact file format matching Sankode's `\n===भार===\n` delimiter.
        """
        os.makedirs(os.path.dirname(os.path.abspath(filepath)), exist_ok=True)

        vocab_str = "".join(self.vocab.tokens)
        weights_str = ",".join(f"{w:.6f}" for w in self.weights)
        payload = f"{vocab_str}\n===भार===\n{weights_str}"

        with open(filepath, "w", encoding="utf-8") as f:
            f.write(payload)

    @classmethod
    def load(cls, filepath: str) -> LanguageModel:
        """
        Loads vocabulary and weights from a serialized weights file on disk.
        """
        with open(filepath, "r", encoding="utf-8") as f:
            content = f.read()

        parts = content.split("\n===भार===\n")
        vocab_str = parts[0]
        weights_str = parts[1]

        tokens = list(vocab_str)
        vocab = Vocabulary(tokens)

        weights = [float(x.strip()) for x in weights_str.split(",") if x.strip()]
        return cls(vocab, weights)

    def generate(
        self,
        rng: RandomGenerator,
        prompt: str,
        length: int = 50,
        temperature: float = 0.70,
    ) -> str:
        """
        Autoregressively generates text from an initial prompt using temperature-scaled sampling.
        """
        if not prompt:
            return ""

        result = prompt
        curr_char = prompt[-1]
        v = self.size

        for _ in range(length):
            u = self.vocab.find(curr_char)
            if u < 0:
                # Unseen character fallback: uniform random sample
                next_idx = rng.next_int(v)
                curr_char = self.vocab.get(next_idx)
                result += curr_char
                continue

            # Compute temperature-scaled unnormalized probabilities: P^(1/T)
            base_idx = u * v
            scaled_exp = [0.0] * v
            exp_sum = 0.0

            for c in range(v):
                prob = self.weights[base_idx + c]
                val = math.exp(math.log(prob) / temperature)
                scaled_exp[c] = val
                exp_sum += val

            # Sample from categorical distribution via cumulative sum
            rand_point = rng.random_float() * exp_sum
            cum_sum = 0.0
            chosen_idx = v - 1

            for c in range(v):
                cum_sum += scaled_exp[c]
                if cum_sum >= rand_point:
                    chosen_idx = c
                    break

            next_char = self.vocab.get(chosen_idx)
            result += next_char
            curr_char = next_char

        return result


# ==============================================================================
# 4. Main Pipeline
# ==============================================================================
def main() -> None:
    print("================================================================")
    print("॥ पायथन-तन्त्रे अभिज्ञानशाकुन्तल-सम्पूर्ण-प्रतिरूपम् ॥")
    print("Kalidasa's Abhijnanasakuntalam LLM Pipeline (Python Reference)")
    print("================================================================")

    repo_root = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", ".."))
    corpus_file = os.path.join(repo_root, "examples", "अभिज्ञानशाकुन्तलम्_मूलम्.पाठ")
    weights_file = os.path.join(repo_root, "examples", "शाकुन्तल_प्रतिरूप_python.भार")

    # Step 1: Read Corpus
    print("--- 1. Reading Corpus File (संचिकातः शाकुन्तल-पाठ-पठनम्) ---")
    if not os.path.exists(corpus_file):
        print(f"Error: Corpus file not found at {corpus_file}")
        sys.exit(1)

    with open(corpus_file, "r", encoding="utf-8") as f:
        corpus = f.read()

    print(f"Corpus Character Length (वर्णदैर्घ्यम्): {len(corpus)}")

    # Step 2: Build Vocabulary
    vocab = Vocabulary()
    for ch in corpus:
        vocab.add(ch)

    print(f"Vocabulary Size (विशिष्ट-वर्णानां सङ्ख्या): {vocab.size}")

    # Step 3: Train Model
    print("--- 2. Training Autoregressive Bigram Model (प्रतिरूपस्य प्रशिक्षणम्) ---")
    model = LanguageModel(vocab)
    model.train(corpus)
    print("Model training complete (प्रतिरूपस्य प्रशिक्षणं सम्पन्नम्)!")

    loss = model.compute_loss(corpus)
    perplexity = math.exp(loss)
    print(f"Cross-Entropy Loss (प्रशिक्षण-हानिः): {loss:.4f}")
    print(f"Perplexity (प्रतिरूपस्य संभ्रमः): {perplexity:.4f}")

    # Step 4: Save Weights
    print("--- 3. Saving Model Weights to Disk (संचिकायां भार-सञ्चयः) ---")
    model.save(weights_file)
    print(f"Weights successfully saved to: {weights_file}")

    # Step 5: Load Weights from Disk
    print("--- 4. Loading Model Weights from Disk (संचिकातः भारोद्धारः) ---")
    loaded_model = LanguageModel.load(weights_file)
    print("Weights successfully loaded from disk!")
    print(f"Loaded Vocabulary Size: {loaded_model.size}")

    # Step 6: Test Generative Inference
    print("--- 5. Testing Generative Inference (भाषा-सृजन-परीक्षणम्) ---")
    rng = RandomGenerator(seed=987654321)

    prompts = ["शकुन्तला", "या सृष्टिः", "दुष्यन्तः"]
    for i, prompt in enumerate(prompts, 1):
        print(f"[Generation {i}: Prompt '{prompt}']:")
        output = loaded_model.generate(rng, prompt=prompt, length=50, temperature=0.70)
        print(output)
        print()

    print("================================================================")
    print("॥ सम्पूर्ण-प्रक्रिया सफलीभूता (Full Pipeline Succeeded!) ॥")
    print("================================================================")


if __name__ == "__main__":
    main()

File: examples/comparative/shakuntala_llm.c
/*
================================================================================
Kalidasa's Abhijnanasakuntalam Autoregressive Language Model (C Reference)
महाकविकालिदासविरचितस्य अभिज्ञानशाकुन्तलस्य भाषा-प्रतिरूपम् (सी-तुलना)
================================================================================

This program is the official, well-documented C99/C11 equivalent of the pure
Devanagari Sankode implementation found in `examples/शाकुन्तल_सम्पूर्ण_प्रतिरूप.सङ्`.

Comparative Architecture & Memory Model Notes:
---------------------------------------------
1. Manual Memory vs. Affine Ownership:
   - In C: All dynamic buffers (`weights`, `corpus`, `tokens`) require explicit
     `malloc()` and `free()`. Failure to free causes memory leaks; dangling
     pointers cause undefined behavior or security exploits.
   - In Sankode: Memory is governed by compile-time affine ownership (`स्वामित्व`).
     When a `संरचना` or `सूची` goes out of scope, it is deterministically freed
     with zero garbage collection pauses and zero use-after-free risks.

2. UTF-8 Devanagari Strings:
   - In C: Strings are raw `char*` byte arrays. Devanagari characters span 3 bytes
     each in UTF-8 (`0xE0 0xA4/0xA5 ...`). We implement a custom UTF-8 character
     boundary decoder (`utf8_char_len`).
   - In Sankode: UTF-8 strings (`सूत्र`) and character tokens are native, first-class
     types with compile-time immutability and built-in intrinsics (`सूत्र_दैर्घ्यम्`,
     `सूत्र_वर्ण`, `सूत्र_विभाजय`).

3. Borrow Checking vs. Raw Pointers:
   - In C: Passing pointers (`LanguageModel*`) offers zero aliasing guarantees.
     Multiple pointers can simultaneously mutate shared state, leading to race conditions.
   - In Sankode: The borrow checker enforces XOR aliasing: either multiple shared
     immutable borrows (`ऋण`), OR exactly one exclusive mutable borrow (`चलऋण`).
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#ifdef _WIN32
#include <windows.h>
#endif

#define MAX_TOKENS 128
#define MAX_TOKEN_BYTES 8
#define SEED_INIT 987654321LL
#define DELIMITER "\n===भार===\n"

/* Cross-platform UTF-8 file open (supporting Unicode Devanagari paths on Windows) */
static FILE* utf8_fopen(const char* filename, const char* mode) {
#ifdef _WIN32
    wchar_t wfilename[1024];
    wchar_t wmode[32];
    MultiByteToWideChar(CP_UTF8, 0, filename, -1, wfilename, 1024);
    MultiByteToWideChar(CP_UTF8, 0, mode, -1, wmode, 32);
    return _wfopen(wfilename, wmode);
#else
    return fopen(filename, mode);
#endif
}


/* =============================================================================
 * 1. UTF-8 Utility Functions (सूत्र-परीक्षण-साधनानि)
 * =============================================================================
 * Devanagari characters in Unicode NFKC are encoded as 3-byte sequences in UTF-8.
 * Standard ASCII characters (spaces, punctuation) are 1 byte.
 */
static int utf8_char_len(unsigned char c) {
    if ((c & 0x80) == 0x00) return 1; /* 0xxxxxxx: ASCII */
    if ((c & 0xE0) == 0xC0) return 2; /* 110xxxxx */
    if ((c & 0xF0) == 0xE0) return 3; /* 1110xxxx: Devanagari letters & matras */
    if ((c & 0xF8) == 0xF0) return 4; /* 11110xxx */
    return 1;
}

/* =============================================================================
 * 2. Pseudo-Random Number Generator (यादृच्छिक-संख्या-यन्त्रम्)
 * =============================================================================
 * 31-bit Linear Congruential Generator matching Sankode's exact parameters.
 */
typedef struct {
    long long seed;
} RandomGenerator;

void rng_init(RandomGenerator* rng, long long seed) {
    rng->seed = seed;
}

int rng_next_int(RandomGenerator* rng, int limit) {
    if (limit <= 1) return 0;
    rng->seed = (rng->seed * 1664525LL + 1013904223LL) % 2147483647LL;
    long long res = rng->seed % limit;
    if (res < 0) res = -res;
    return (int)res;
}

double rng_random_float(RandomGenerator* rng) {
    int val = rng_next_int(rng, 10000);
    return (double)val / 10000.0;
}

/* =============================================================================
 * 3. Vocabulary & Character Tokenizer (शब्दावली)
 * =============================================================================
 */
typedef struct {
    char tokens[MAX_TOKENS][MAX_TOKEN_BYTES];
    int size;
} Vocabulary;

void vocab_init(Vocabulary* v) {
    v->size = 0;
    memset(v->tokens, 0, sizeof(v->tokens));
}

int vocab_find(const Vocabulary* v, const char* token) {
    for (int i = 0; i < v->size; i++) {
        if (strcmp(v->tokens[i], token) == 0) {
            return i;
        }
    }
    return -1;
}

int vocab_add(Vocabulary* v, const char* token) {
    int existing = vocab_find(v, token);
    if (existing >= 0) return existing;
    if (v->size >= MAX_TOKENS) return -1;

    strncpy(v->tokens[v->size], token, MAX_TOKEN_BYTES - 1);
    v->tokens[v->size][MAX_TOKEN_BYTES - 1] = '\0';
    v->size++;
    return v->size - 1;
}

const char* vocab_get(const Vocabulary* v, int idx) {
    if (idx >= 0 && idx < v->size) {
        return v->tokens[idx];
    }
    return "";
}

/* =============================================================================
 * 4. Autoregressive Language Model (भाषा-प्रतिरूपम्)
 * =============================================================================
 */
typedef struct {
    Vocabulary vocab;
    int size;          /* |V| */
    double* weights;   /* Flattened 2D transition probability matrix of size |V| * |V| */
} LanguageModel;

void model_init(LanguageModel* m, const Vocabulary* v) {
    m->vocab = *v;
    m->size = v->size;
    int total = m->size * m->size;
    m->weights = (double*)calloc(total > 0 ? total : 1, sizeof(double));
}

void model_free(LanguageModel* m) {
    if (m->weights) {
        free(m->weights);
        m->weights = NULL;
    }
    m->size = 0;
}

/*
 * Trains the autoregressive bigram transition model on UTF-8 corpus text.
 * Uses Laplace add-one smoothing to prevent zero-probability transitions.
 */
void model_train(LanguageModel* m, const char* corpus) {
    int v = m->size;
    int total = v * v;
    double* counts = (double*)calloc(total, sizeof(double));
    double* row_sums = (double*)calloc(v, sizeof(double));

    const unsigned char* ptr = (const unsigned char*)corpus;
    char prev_tok[MAX_TOKEN_BYTES] = {0};
    int has_prev = 0;

    while (*ptr) {
        int len = utf8_char_len(*ptr);
        char curr_tok[MAX_TOKEN_BYTES] = {0};
        memcpy(curr_tok, ptr, len);
        curr_tok[len] = '\0';

        if (has_prev) {
            int u = vocab_find(&m->vocab, prev_tok);
            int w = vocab_find(&m->vocab, curr_tok);
            if (u >= 0 && w >= 0) {
                counts[u * v + w] += 1.0;
                row_sums[u] += 1.0;
            }
        }

        memcpy(prev_tok, curr_tok, MAX_TOKEN_BYTES);
        has_prev = 1;
        ptr += len;
    }

    /* Laplace Smoothing & Normalization */
    for (int r = 0; r < v; r++) {
        double denom = row_sums[r] + (double)v;
        for (int c = 0; c < v; c++) {
            int idx = r * v + c;
            m->weights[idx] = (counts[idx] + 1.0) / denom;
        }
    }

    free(counts);
    free(row_sums);
}

/*
 * Computes cross-entropy loss (Negative Log-Likelihood) over the corpus.
 */
double model_compute_loss(const LanguageModel* m, const char* corpus) {
    int v = m->size;
    const unsigned char* ptr = (const unsigned char*)corpus;
    char prev_tok[MAX_TOKEN_BYTES] = {0};
    int has_prev = 0;

    double total_nll = 0.0;
    double count = 0.0;

    while (*ptr) {
        int len = utf8_char_len(*ptr);
        char curr_tok[MAX_TOKEN_BYTES] = {0};
        memcpy(curr_tok, ptr, len);
        curr_tok[len] = '\0';

        if (has_prev) {
            int u = vocab_find(&m->vocab, prev_tok);
            int w = vocab_find(&m->vocab, curr_tok);
            if (u >= 0 && w >= 0) {
                double prob = m->weights[u * v + w];
                if (prob > 0.0) {
                    total_nll -= log(prob);
                    count += 1.0;
                }
            }
        }

        memcpy(prev_tok, curr_tok, MAX_TOKEN_BYTES);
        has_prev = 1;
        ptr += len;
    }

    return count > 0.0 ? total_nll / count : 0.0;
}

/*
 * Saves vocabulary and weights to disk in format matching Sankode.
 */
int model_save(const LanguageModel* m, const char* filepath) {
    FILE* f = utf8_fopen(filepath, "wb");
    if (!f) return 0;

    /* Write vocabulary characters */
    for (int i = 0; i < m->vocab.size; i++) {
        fputs(m->vocab.tokens[i], f);
    }
    fputs(DELIMITER, f);

    /* Write comma-separated weights */
    int total = m->size * m->size;
    for (int i = 0; i < total; i++) {
        if (i > 0) fputc(',', f);
        fprintf(f, "%.6f", m->weights[i]);
    }

    fclose(f);
    return 1;
}

/*
 * Loads vocabulary and weights from disk.
 */
int model_load(LanguageModel* m, const char* filepath) {
    FILE* f = utf8_fopen(filepath, "rb");
    if (!f) return 0;

    fseek(f, 0, SEEK_END);
    long fsize = ftell(f);
    fseek(f, 0, SEEK_SET);

    char* buffer = (char*)malloc(fsize + 1);
    fread(buffer, 1, fsize, f);
    buffer[fsize] = '\0';
    fclose(f);

    char* delim_pos = strstr(buffer, DELIMITER);
    if (!delim_pos) {
        free(buffer);
        return 0;
    }

    *delim_pos = '\0';
    char* vocab_part = buffer;
    char* weights_part = delim_pos + strlen(DELIMITER);

    /* Reconstruct vocabulary */
    Vocabulary v;
    vocab_init(&v);
    unsigned char* vptr = (unsigned char*)vocab_part;
    while (*vptr) {
        int len = utf8_char_len(*vptr);
        char tok[MAX_TOKEN_BYTES] = {0};
        memcpy(tok, vptr, len);
        tok[len] = '\0';
        vocab_add(&v, tok);
        vptr += len;
    }

    model_init(m, &v);

    /* Parse comma-separated weights */
    int total = m->size * m->size;
    char* token = strtok(weights_part, ",\r\n");
    int idx = 0;
    while (token && idx < total) {
        m->weights[idx++] = atof(token);
        token = strtok(NULL, ",\r\n");
    }

    free(buffer);
    return 1;
}

/*
 * Generates text from a prompt using temperature-controlled softmax sampling.
 */
void model_generate(
    LanguageModel* m,
    RandomGenerator* rng,
    const char* prompt,
    int length,
    double temperature,
    char* out_buf,
    size_t out_buf_size
) {
    if (!prompt || strlen(prompt) == 0) {
        out_buf[0] = '\0';
        return;
    }

    strcpy(out_buf, prompt);

    /* Extract the last UTF-8 character from prompt */
    int p_len = (int)strlen(prompt);
    int last_char_start = 0;
    int cur = 0;
    while (cur < p_len) {
        last_char_start = cur;
        cur += utf8_char_len((unsigned char)prompt[cur]);
    }
    char curr_tok[MAX_TOKEN_BYTES] = {0};
    strncpy(curr_tok, prompt + last_char_start, cur - last_char_start);
    curr_tok[cur - last_char_start] = '\0';

    int v = m->size;
    double* scaled_exp = (double*)malloc(v * sizeof(double));

    for (int step = 0; step < length; step++) {
        int u = vocab_find(&m->vocab, curr_tok);
        if (u < 0) {
            int next_idx = rng_next_int(rng, v);
            const char* next_tok = vocab_get(&m->vocab, next_idx);
            strncat(out_buf, next_tok, out_buf_size - strlen(out_buf) - 1);
            strncpy(curr_tok, next_tok, MAX_TOKEN_BYTES);
            continue;
        }

        int base_idx = u * v;
        double exp_sum = 0.0;
        for (int c = 0; c < v; c++) {
            double prob = m->weights[base_idx + c];
            double val = exp(log(prob) / temperature);
            scaled_exp[c] = val;
            exp_sum += val;
        }

        double rand_point = rng_random_float(rng) * exp_sum;
        double cum_sum = 0.0;
        int chosen_idx = v - 1;

        for (int c = 0; c < v; c++) {
            cum_sum += scaled_exp[c];
            if (cum_sum >= rand_point) {
                chosen_idx = c;
                break;
            }
        }

        const char* next_tok = vocab_get(&m->vocab, chosen_idx);
        strncat(out_buf, next_tok, out_buf_size - strlen(out_buf) - 1);
        strncpy(curr_tok, next_tok, MAX_TOKEN_BYTES);
    }

    free(scaled_exp);
}

/* =============================================================================
 * 5. Main Pipeline
 * =============================================================================
 */
int main(void) {
#ifdef _WIN32
    SetConsoleOutputCP(CP_UTF8);
#endif

    printf("================================================================\n");
    printf("॥ सी-तन्त्रे अभिज्ञानशाकुन्तल-सम्पूर्ण-प्रतिरूपम् ॥\n");
    printf("Kalidasa's Abhijnanasakuntalam LLM Pipeline (C Reference)\n");
    printf("================================================================\n");

    const char* corpus_file = "examples/अभिज्ञानशाकुन्तलम्_मूलम्.पाठ";
    const char* weights_file = "examples/शाकुन्तल_प्रतिरूप_c.भार";

    /* Step 1: Read Corpus */
    printf("--- 1. Reading Corpus File (संचिकातः शाकुन्तल-पाठ-पठनम्) ---\n");
    FILE* f = utf8_fopen(corpus_file, "rb");
    if (!f) {
        fprintf(stderr, "Error opening corpus file: %s\n", corpus_file);
        return 1;
    }
    fseek(f, 0, SEEK_END);
    long fsize = ftell(f);
    fseek(f, 0, SEEK_SET);

    char* corpus = (char*)malloc(fsize + 1);
    fread(corpus, 1, fsize, f);
    corpus[fsize] = '\0';
    fclose(f);

    /* Normalize CRLF to LF */
    char* wptr = corpus;
    for (char* rptr = corpus; *rptr; rptr++) {
        if (*rptr != '\r') {
            *wptr++ = *rptr;
        }
    }
    *wptr = '\0';

    /* Count UTF-8 characters */
    int char_count = 0;
    unsigned char* ptr = (unsigned char*)corpus;
    while (*ptr) {
        ptr += utf8_char_len(*ptr);
        char_count++;
    }
    printf("Corpus Character Length (वर्णदैर्घ्यम्): %d\n", char_count);

    /* Step 2: Build Vocabulary */
    Vocabulary vocab;
    vocab_init(&vocab);
    ptr = (unsigned char*)corpus;
    while (*ptr) {
        int len = utf8_char_len(*ptr);
        char tok[MAX_TOKEN_BYTES] = {0};
        memcpy(tok, ptr, len);
        tok[len] = '\0';
        vocab_add(&vocab, tok);
        ptr += len;
    }
    printf("Vocabulary Size (विशिष्ट-वर्णानां सङ्ख्या): %d\n", vocab.size);

    /* Step 3: Train Model */
    printf("--- 2. Training Autoregressive Bigram Model (प्रतिरूपस्य प्रशिक्षणम्) ---\n");
    LanguageModel model;
    model_init(&model, &vocab);
    model_train(&model, corpus);
    printf("Model training complete (प्रतिरूपस्य प्रशिक्षणं सम्पन्नम्)!\n");

    double loss = model_compute_loss(&model, corpus);
    double perplexity = exp(loss);
    printf("Cross-Entropy Loss (प्रशिक्षण-हानिः): %.4f\n", loss);
    printf("Perplexity (प्रतिरूपस्य संभ्रमः): %.4f\n", perplexity);

    /* Step 4: Save Weights */
    printf("--- 3. Saving Model Weights to Disk (संचिकायां भार-सञ्चयः) ---\n");
    if (model_save(&model, weights_file)) {
        printf("Weights successfully saved to: %s\n", weights_file);
    } else {
        fprintf(stderr, "Failed to save weights to: %s\n", weights_file);
    }

    /* Step 5: Load Weights */
    printf("--- 4. Loading Model Weights from Disk (संचिकातः भारोद्धारः) ---\n");
    LanguageModel loaded_model;
    if (model_load(&loaded_model, weights_file)) {
        printf("Weights successfully loaded from disk!\n");
        printf("Loaded Vocabulary Size: %d\n", loaded_model.size);
    } else {
        fprintf(stderr, "Failed to load weights from: %s\n", weights_file);
        return 1;
    }

    /* Step 6: Test Generative Inference */
    printf("--- 5. Testing Generative Inference (भाषा-सृजन-परीक्षणम्) ---\n");
    RandomGenerator rng;
    rng_init(&rng, SEED_INIT);

    const char* prompts[] = {"शकुन्तला", "या सृष्टिः", "दुष्यन्तः"};
    char gen_buf[2048];

    for (int i = 0; i < 3; i++) {
        printf("[Generation %d: Prompt '%s']:\n", i + 1, prompts[i]);
        model_generate(&loaded_model, &rng, prompts[i], 50, 0.70, gen_buf, sizeof(gen_buf));
        printf("%s\n\n", gen_buf);
    }

    printf("================================================================\n");
    printf("॥ सम्पूर्ण-प्रक्रिया सफलीभूता (Full Pipeline Succeeded!) ॥\n");
    printf("================================================================\n");

    /* Free all allocated memory */
    free(corpus);
    model_free(&model);
    model_free(&loaded_model);

    return 0;
}